### PR TITLE
chore(tests): Get benchmarks running again

### PR DESCRIPTION
**Pull request by jszwedko**
The commits in this PR can be viewed in order.

The high-level changes are:

* Fixes WASM and regular benchmarks so they run again
* Adds `--all-targets` to `make test` and `make test-wasm` so that the benchmarks are also tested. Currently this is kinda slow as we have many benchmarks where a single sample takes many seconds, but I intend to improve this by removing the internal batching from the samples where it isn't necessary. I was originally going to add `make` target specifically to run the benchmarks as tests, but I couldn't find way to do this (`cargo test --benches` runs the unit tests too; `cargo test --benches -- --bench` skips the unit tests but runs the benchmarks as benchmarks and not tests).
* Adds the WASM tests and benchmark tests to CI; these currently don't seem to be have run in CI anywhere

Closes: #4284

### CODE REVIEW

File: benches/lua.rs
Comment (Hoverbear) [inline benches/lua.rs:87]: How would you feel about making this per-event?

File: benches/isolated_buffering.rs
Comment (Hoverbear) [inline benches/isolated_buffering.rs:31]: How would you feel about reducing this batch size? It's a lot of events since criterion will run this in a loop with warmup.

File: benches/files.rs
Comment (Hoverbear) [inline benches/files.rs:15]: Similarly, how would you feel about reducing this down? Criterion is already running in a fairly hot loop and this is a very coarse grained measurement.

File: benches/buffering.rs
Comment (Hoverbear) [inline benches/buffering.rs:11]: I feel like I'm repeating myself, but this is a very coarse grained measurement and criterion is already doing several thousand iterations for us.

File: benches/bench.rs
Comment (Hoverbear) [inline benches/bench.rs:464]: "Ana repeats herself about coarse grained measurements"

File: benches/bench.rs
Comment (Hoverbear) [inline benches/bench.rs:634]: Could you use `debug_assert` for the test runs, so that they don't affect measurement in release? https://doc.rust-lang.org/std/macro.debug_assert.html

File: benches/batch.rs
Comment (Hoverbear) [inline benches/batch.rs:92]: Below, in this file, theyre are a bench of benches calling this code that use high `num_events` numbers, and I repeat myself about coarse grained samples and criterion already doing this.

**Review by Hoverbear — COMMENTED** [review]
Really great work so far. :) I'm fine merging this, but I would advise we either reduce our self-set batches and rely on Criterion's facilities a bit more. Maybe i'ts best to plan some future work for that though.

File: benches/lua.rs
Comment (jszwedko) [inline benches/lua.rs:25]: Switched away from `b.iter_with_setup` as:

* It's soft deprecated
* It was running this once per iteration which seemed unnecessary and possibly working against the purpose of the benchmark to measure the throughput

File: benches/lua.rs
Comment (jszwedko) [inline benches/lua.rs:22]: Switched away from `c.bench` to creating a benchmark group given `c.bench` is deprecated: https://docs.rs/criterion/0.3.3/src/criterion/lib.rs.html#1634

File: benches/lua.rs
Comment (jszwedko) [reply benches/lua.rs:65]: I did switch this to `iter_batched`, but I think this one still fits better running a small batch given we want to, sometimes, filter an event (10% of the time). I could augment it to be an individual event that is randomly filtered or not filtered, but I figured there could be some difference in execution time for a filtered event vs. an unfiltered event.

I could change this to be two benchmarks for the filter though, one where it always filters and one where it never filters.

File: benches/lua.rs
Comment (jszwedko) [inline benches/lua.rs:20]: The body of the iter loop for each benchmark was the same so I refactored this to be an array of (name, transform) tuples. I do get the feeling that this could be a little overboard though :smile: .

File: benches/lua.rs
Comment (jszwedko) [inline benches/lua.rs:55]: I added these `black_box` to ensure that the compiler doesn't optimize away the benchmark due to it discovering that the value is never actually used (except in debug mode where there is an assertion).

criterion automatically wraps inputs when `bench_with_input` is used and appears to wrap outputs of the iter closure. Instead of `black_box` I could write this as:

```
                 let event = transform.transform(event.clone()).unwrap();
                 debug_assert_eq!(event.as_log()[key], value.to_owned().into());
                 event
```

To return the event and have criterion wrap it in `black_box`.

File: benches/lua.rs
Comment (jszwedko) [inline benches/lua.rs:11]: It might be easiest to view this diff side-by-side.

File: benches/lua.rs
Comment (Hoverbear) [reply benches/lua.rs:20]: I like it! You might prefer to use `AddFields::from(toml::from_str("my config block"))` for consistency's sake.

File: benches/lua.rs
Comment (Hoverbear) [inline benches/lua.rs:78]: Oh no rustfmt :(

**Comment by jszwedko** [comment]
@Hoverbear I have some merge conflicts to resolve tomorrow, but I think this is ready for review in the meanwhile. I recommend checking out the last 4 commits. https://github.com/timberio/vector/pull/4809/commits/92aed118508201bb5e8317ab7478584f577805a8 is the big one. It was mostly mechanical, but resulted in a rather large diff. I can try to split it up if needed.

I don't think the result is perfect, but I do think it is an improvement. I'd still like to see the large end-to-end benchmarks improved (or possibly removed if they are not valuable).

I was hoping to get your thoughts on testing the WASM benchmarks separately too (https://github.com/timberio/vector/pull/4809/commits/9ee9aee51d6216ad46a8332694c9a05ac082cc0e). It's a bit awkward as it requires a few features aside from the `transforms-wasm` itself. I'm almost wondering if it'd be better to do one build that has WASM _and_ all the default features for testing rather than separating them out (maybe a `test-with-wasm` rather than `test-wasm`).

**Comment by Hoverbear** [comment]
> It's a bit awkward as it requires a few features aside from the transforms-wasm itself. I'm almost wondering if it'd be better to do one build that has WASM and all the default features for testing rather than separating them out (maybe a test-with-wasm rather than test-wasm).

That totally works for me!

**Comment by Hoverbear** [comment]
Noting that this does not appear to compile yet:

```bash
ana@autonoma:~/git/timberio/vector$ git log -p
commit 9ee9aee51d6216ad46a8332694c9a05ac082cc0e (HEAD -> fix-benchmarks, origin/fix-benchmarks)
Author: Jesse Szwedko <jesse@szwedko.me>
Date:   Wed Nov 4 22:21:25 2020 +0000
# ...
ana@autonoma:~/git/timberio/vector$ git pull
Already up to date.
ana@autonoma:~/git/timberio/vector$ make bench
cargo bench --no-default-features --features "default" | tee bench_output.txt
   Compiling vector v0.11.0 (/home/ana/git/timberio/vector)
error[E0601]: `main` function not found in crate `http`
   --> benches/http.rs:1:1
    |
1   | / use criterion::{criterion_group, BatchSize, BenchmarkId, Criterion, Throughput};
2   | | use futures::{compat::Future01CompatExt, TryFutureExt};
3   | | use hyper::{
4   | |     service::{make_service_fn, service_fn},
...   |
102 | |
103 | | criterion_group!(http, benchmark_http);
    | |_______________________________________^ consider adding a `main` function to `benches/http.rs`

error[E0601]: `main` function not found in crate `event`
  --> benches/event.rs:1:1
   |
1  | / use bytes::Bytes;
2  | | use criterion::{criterion_group, BatchSize, Criterion};
3  | | use serde_json::{json, Value};
4  | | use vector::{
...  |
98 | |
99 | | criterion_group!(event, benchmark_event);
   | |_________________________________________^ consider adding a `main` function to `benches/event.rs`

error[E0601]: `main` function not found in crate `files`
  --> benches/files.rs:1:1
   |
1  | / use bytes::Bytes;
2  | | use criterion::{criterion_group, BatchSize, Criterion, Throughput};
3  | | use futures::{compat::Future01CompatExt, stream, SinkExt, StreamExt};
4  | | use std::convert::TryInto;
...  |
93 | |
94 | | criterion_group!(files, benchmark_files_without_partitions);
   | |____________________________________________________________^ consider adding a `main` function to `benches/files.rs`

warning: unused import: `approx::assert_relative_eq`
 --> benches/bench.rs:3:5
  |
3 | use approx::assert_relative_eq;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0601]: `main` function not found in crate `batch`
   --> benches/batch.rs:1:1
    |
1   | / use bytes::Bytes;
2   | | use criterion::{criterion_group, Criterion, Throughput};
3   | | use futures::{compat::Future01CompatExt, future};
4   | | use futures01::{Sink, Stream};
...   |
174 | |     }
175 | | }
    | |_^ consider adding a `main` function to `benches/batch.rs`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0601`.
error: could not compile `vector`.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: aborting due to previous error

For more information about this error, try `rustc --explain E0601`.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0601`.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0601`.
```

**Comment by jszwedko** [comment]
Doh, I must have broken it. I'll resolve it, the merge conflicts, and add the `test-with-wasm` `make` target.

**Comment by jszwedko** [comment]
Noting that this should be working now, but I commented out:

* The WASM benchmarks given the transform needs to be updated
* The Lua benchmarks pending https://github.com/timberio/vector/pull/4875 as Ana and I could not satisfy the borrow checker. We believe the changes in that PR should make it easier (specifically the addition of `Sync` to `FunctionTransform`).

**Comment by Hoverbear** [comment]
Confirming #4875 will unblock the lua benches.

File: Makefile
Comment (Hoverbear) [inline Makefile:807]: Probably better to put this after the normal benches.

File: Makefile
Comment (Hoverbear) [inline Makefile:808]: If you make these two commands use the same feature set you can probably skip a long compile time.

**Comment by Hoverbear** [comment]
@jszwedko I spotted this error while testing locally: 
image

**Review by Hoverbear — APPROVED** [review]
Oh yes I'm so excited to be able to use these again. :) I think other than the issue I raised it looks good.

This is great:

image

File: Makefile
Comment (jszwedko) [reply Makefile:808]: :+1: I mean to replace 808 with 807 rather than having both in there

**Comment by jszwedko** [comment]
Just noting that https://github.com/timberio/vector/pull/4875 is required for renabling the wasm benchmarks too for similar reasons to the lua one where we are benchmarking both `FunctionTransform` and `TaskTransform` together.

**Comment by Hoverbear** [comment]
@jszwedko I think that PR will merge soon but if you want we could merge this first and I can repair them in that PR.

File: Makefile
Comment (Hoverbear) [inline Makefile:278]: It seems our naming convention suggests `test-components` here.

**Comment by Hoverbear** [comment]
@jszwedko I think my final review comment before approval is to ask if we must have `benches/default_benchmarks` and `benches/wasm_benchmarks`... Could we have `benches/` and `benches/wasm/`?

**Comment by jszwedko** [comment]
Merged in #4875 and updated lua and wasm transforms. I'll wait for #4875 to merge before merging this one.

> @jszwedko I think my final review comment before approval is to ask if we must have `benches/default_benchmarks` and `benches/wasm_benchmarks`... Could we have `benches/` and `benches/wasm/`?

:+1: I'm fine with that.

**Comment by jszwedko** [comment]
Moved the default benchmarks back up.

I'll double check this on Monday, but I think it should be good to go.

File: benches/wasm/mod.rs
Comment (jszwedko) [inline benches/wasm/mod.rs]: @Hoverbear I'm curious if you have any thoughts here. This benchmark currently hangs forever as `bench_group_transforms_over_parameterized_event_sizes` determines when an event has been processed by polling the output of the transform (which in this case, no event is emitted). I could not determine a good way to otherwise determine when an event has been fully processed by a transform.

File: Makefile
Comment (Hoverbear) [inline Makefile:858]: Why this? Criterion generates rich HTML reports in `target/criterion` no?

Comment (jszwedko) [reply Makefile:858]: Do you mean the `tee`? That was already there as part of https://github.com/timberio/vector/pull/4630

I agree though, I'll revisit this as part of #4758

Comment (Hoverbear) [reply Makefile:858]: Oh! Ok! I missed this!

File: benches/wasm/mod.rs
Comment (Hoverbear) [reply benches/wasm/mod.rs]: Ohhhh, I think we need to just drop this for now then. :( With the task transform changes I think the only assertion we can make is that "nothing ever arrives" which seems to be passing right now.

**Comment by jszwedko** [comment]
@Hoverbear I think this is _finally_ ready to be merged.

**Review by Hoverbear — APPROVED** [review]
Rest looks good :)

File: .github/workflows/master.yml
Comment (jszwedko) [reply .github/workflows/master.yml]: Oops, yep, thank you :smile:

File: Makefile
Comment (jszwedko) [reply Makefile:275]: Hmm, would I not want to run those tests too though?

**Pull request merged**
